packages/pluggableWidgets/bar-chart-native/src/utils/SeriesLoader.ts

```diff
@@ -0,0 +1,230 @@
+                return group.groupByAttributeValue.eq(groupByAttributeValue.value);
+            }
+
+            return group.groupByAttributeValue === groupByAttributeValue.value;
+        });
+
+        if (group) {
+            group.items.push(item);
+        } else {
+            const newDataSourceItemGroup: DataSourceItemGroup = {
+                groupByAttributeValue: groupByAttributeValue.value,
+                items: [item]
+            };
+
+            if (dynamicSeriesName) {
+                const dynamicSeriesNameValue = dynamicSeriesName(item);
+
+                if (dynamicSeriesNameValue.value === undefined) {
+                    return null;
+                }
+
+                newDataSourceItemGroup.dynamicSeriesNameValue = dynamicSeriesNameValue.value;
+            }
+
+            if (dynamicCustomBarStyle) {
+                const dynamicCustomBarStyleValue = dynamicCustomBarStyle(item);
+
+                if (dynamicCustomBarStyleValue.value === undefined) {
+                    return null;
+                }
+
+                newDataSourceItemGroup.dynamicCustomBarStyleValue = dynamicCustomBarStyleValue.value;
+            }
+
+            dataSourceItemGroupsResult.push(newDataSourceItemGroup);
+        }
+    }
+
+    return dataSourceItemGroupsResult;
+}
+
+function extractDataPoints(series: BarSeriesType, dataSourceItems?: ObjectItem[]): DataPointsExtraction | null {
+    if (!dataSourceItems) {
+        // todo: why is this check here? dataSourceItems are only passed for dynamic items.
+        const dataSource =
+            series.dataSet === "static" ? ensure(series.staticDataSource) : ensure(series.dynamicDataSource);
+
+        if (!dataSource.items) {
+            return null;
+        }
+
+        dataSourceItems = dataSource.items;
+    }
+
+    const dataPointsExtraction: DataPointsExtraction = { dataPoints: [] };
+
+    for (const item of dataSourceItems) {
+        const x = (series.dataSet === "static" ? ensure(series.staticXAttribute) : ensure(series.dynamicXAttribute))(
+            item
+        );
+        const y = (series.dataSet === "static" ? ensure(series.staticYAttribute) : ensure(series.dynamicYAttribute))(
+            item
+        );
+
+        if (!x.value || !y.value) {
+            return null;
+        }
+
+        if (!dataPointsExtraction.xFormatter && !dataPointsExtraction.yFormatter) {
+            dataPointsExtraction.xFormatter = (xValue: number | Date | string) =>
+                x.formatter.format(typeof xValue === "number" ? new Big(xValue) : xValue);
+            dataPointsExtraction.yFormatter = (yValue: number | Date | string) =>
+                y.formatter.format(typeof yValue === "number" ? new Big(yValue) : yValue);
+        }
+
+        dataPointsExtraction.dataPoints.push({
+            // VictoryChart plots the dependent point (y) on the x axis. In order to align user configuration with
```

Comment:
As this is specific for bar chart, I think it's better to flip this outside the series loader for reusability. It does mean you need to iterate through the series again, which isn't so nice.

Reply:
I agree that the chart-library specific logic should be moved closer to the chart or within it's reponsibility, however based on how the Series Loader has mostly been duplicated, I feel this should be done at a later stage, like when we develop the next chart (pie chart?). We should invest effort into refactoring the Series Loader to handle variances between charts, supporting different attribute types for example, and this would be a good opportunity to cater to "flipping" logic, such as a callback passed to the seriesLoader in the scope of the chart.